File: 0001-0500/0013/0013.roman-to-integer.py
#
# @lc app=leetcode id=13 lang=python3
#
# [13] Roman to Integer
#

# @lc code=start
class Solution:
    def romanToInt(self, s: str) -> int:
        Num = 0
        Roman = {
            'I': 1,
            'V': 5,
            'X': 10,
            'L': 50,
            'C': 100,
            'D': 500,
            'M': 1000,
        }
        Prev = 0
        for letter in s:
            Next = Roman[letter]
            if Prev >= Next:
                Num += Prev
            else:
                Num -= Prev
            Prev = Next
        Num += Next

        return Num
# @lc code=end

File: 0001-0500/0013/0013.roman-to-integer.cpp
/*
 * @lc app=leetcode id=13 lang=cpp
 *
 * [13] Roman to Integer
 */

// @lc code=start
class Solution
{
public:
    int romanToInt(string s)
    {
        int res = 0;
        int len = s.length();
        for (int i = 0; i < len; i++)
        {
            if (i < len - 1 && valueInt(s[i]) < valueInt(s[i + 1]))
            {
                res -= valueInt(s[i]);
            }
            else
            {
                res += valueInt(s[i]);
            }
        }
        return res;
    }
    int valueInt(char c)
    {
        switch (c)
        {
        case 'I':
            return 1;
        case 'V':
            return 5;
        case 'X':
            return 10;
        case 'L':
            return 50;
        case 'C':
            return 100;
        case 'D':
            return 500;
        case 'M':
            return 1000;
        default:
            return 0;
        }
    }
};
// @lc code=end
